### PR TITLE
Remove references to EVP_sha1() and EVP_sha256()

### LINKED ISSUE #14387: Equality tests on EVP_sha1() etc should not be called
In a number of places in the code we have

`if (md == EVP_sha1())`

and that is something we should not be doing. If we want to know if something is SHA1 then testing this way is simply not the right way to do it. It will break in unexpected ways.

```
./ts/ts_rsp_sign.c:        || ctx->ess_cert_id_digest == EVP_sha1()) {
./x509/x_all.c:    if (type == EVP_sha1() && (data->flags & EXFLAG_SET) != 0
./cms/cms_sd.c:            if (md == EVP_sha1() || md == NULL) {
```

### PR DESCRIPTION
**Pull request by paulidale**
There are a few remaining references.  These take two forms:

1. legacy APIs that don't know to free the EVP_MD after they finish using it
2. APIs that only use the name or NID of the EVP_MD

Fixes #14387

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/evp/p5_crpt2.c [hunk 1]
```diff
@@ -21,15 +21,14 @@
 
 int ossl_pkcs5_pbkdf2_hmac_ex(const char *pass, int passlen,
                               const unsigned char *salt, int saltlen, int iter,
-                              const EVP_MD *digest, int keylen,
+                              const char *mdname, int keylen,
```
Comment (slontis) [inline crypto/evp/p5_crpt2.c:24]: I wonder if this change steps on @jon-oracle changes in p12?

File: crypto/ocsp/ocsp_vfy.c [hunk 2]
```diff
@@ -204,8 +205,13 @@ static X509 *ocsp_find_signer_sk(STACK_OF(X509) *certs, OCSP_RESPID *id)
     /* Calculate hash of each key and compare */
     for (i = 0; i < sk_X509_num(certs); i++) {
         x = sk_X509_value(certs, i);
-        if (!X509_pubkey_digest(x, EVP_sha1(), tmphash, NULL))
+        if ((md = EVP_MD_fetch(x->libctx, SN_sha1, x->propq)) == NULL)
             break;
+        if (!X509_pubkey_digest(x, md, tmphash, NULL)) {
```
Comment (slontis) [inline crypto/ocsp/ocsp_vfy.c:210]: sometimes I do the following for these:

````
rv = X509_pubkey_digest(....);
EVP_MD_free(md);
if (!rv)
   break;
````

File: crypto/ocsp/ocsp_vfy.c [hunk 3]
```diff
@@ -204,8 +205,13 @@ static X509 *ocsp_find_signer_sk(STACK_OF(X509) *certs, OCSP_RESPID *id)
     /* Calculate hash of each key and compare */
     for (i = 0; i < sk_X509_num(certs); i++) {
         x = sk_X509_value(certs, i);
-        if (!X509_pubkey_digest(x, EVP_sha1(), tmphash, NULL))
+        if ((md = EVP_MD_fetch(x->libctx, SN_sha1, x->propq)) == NULL)
```
Comment (slontis) [inline crypto/ocsp/ocsp_vfy.c:208]: hopefully x cant be NULL here..

File: crypto/pem/pvkfmt.c [hunk 4]
```diff
@@ -795,16 +795,19 @@ static int derive_pvk_key(unsigned char *key,
                           const unsigned char *pass, int passlen)
 {
     EVP_MD_CTX *mctx = EVP_MD_CTX_new();
+    EVP_MD *md = EVP_MD_fetch(NULL, SN_sha1, NULL);
```
Comment (slontis) [inline crypto/pem/pvkfmt.c:798]: Is this method legacy only? (i.e no libctx)

File: crypto/srp/srp_vfy.c [hunk 5]
```diff
@@ -574,21 +575,27 @@ SRP_user_pwd *SRP_VBASE_get1_by_user(SRP_VBASE *vb, char *username)
 
     if (RAND_priv_bytes(digv, SHA_DIGEST_LENGTH) <= 0)
```
Comment (slontis) [inline crypto/srp/srp_vfy.c:576]: another case where we have multiple things with no libctx.. (and RAND_priv_bytes)

File: crypto/x509/t_x509.c [hunk 6]
```diff
@@ -228,6 +228,7 @@ int X509_ocspid_print(BIO *bp, X509 *x)
     unsigned char SHA1md[SHA_DIGEST_LENGTH];
     ASN1_BIT_STRING *keybstr;
     const X509_NAME *subj;
+    EVP_MD *md = NULL;
 
     /*
```
Comment (slontis) [inline crypto/x509/t_x509.c:233]: maybe check for x == NULL ?

File: providers/implementations/kdfs/scrypt.c [hunk 7]
```diff
@@ -36,7 +36,7 @@ static OSSL_FUNC_kdf_get_ctx_params_fn kdf_scrypt_get_ctx_params;
 static int scrypt_alg(const char *pass, size_t passlen,
                       const unsigned char *salt, size_t saltlen,
                       uint64_t N, uint64_t r, uint64_t p, uint64_t maxmem,
-                      unsigned char *key, size_t keylen, EVP_MD *sha256,
+                      unsigned char *key, size_t keylen, const char *sha256,
```
Comment (slontis) [inline providers/implementations/kdfs/scrypt.c:39]: Why was this change needed? Is prefetching a problem?

File: crypto/ocsp/ocsp_vfy.c (hunk 2 above)
Comment (paulidale) [reply crypto/ocsp/ocsp_vfy.c:210]: Yeah, thought about it.  Didn't bother.

File: crypto/ocsp/ocsp_vfy.c (hunk 3 above)
Comment (paulidale) [reply crypto/ocsp/ocsp_vfy.c:208]: If it is, it crashes on the next step regardless.  Happy to add a NULL check though.

File: crypto/pem/pvkfmt.c (hunk 4 above)
Comment (paulidale) [reply crypto/pem/pvkfmt.c:798]: I've no idea :(

File: providers/implementations/kdfs/scrypt.c (hunk 7 above)
Comment (paulidale) [reply providers/implementations/kdfs/scrypt.c:39]: Instead of passing an EVP_MD, I'm passing the name.  Deep inside, it converts the EVP_MD to its name to fetch.

File: crypto/srp/srp_vfy.c (hunk 5 above)
Comment (paulidale) [reply crypto/srp/srp_vfy.c:576]: `RAND_priv_bytes` doesn't need a libctx, the DRBGs are created once and then it is per thread.

Comment (slontis) [reply crypto/srp/srp_vfy.c:576]: The point being that there are not too many instances of  RAND_priv_bytes() left..
i.e RAND_priv_bytes_ex() is used in most places.

Comment (t8m) [reply crypto/srp/srp_vfy.c:576]: SRP interfaces are deprecated and we should not add _ex() calls to them.

File: crypto/ocsp/ocsp_vfy.c (hunk 3 above)
Comment (paulidale) [reply crypto/ocsp/ocsp_vfy.c:208]: It's coming from a stack of pointers.  If somebody pushes a NULL onto the stack, it's possible.  This seems like a very unlikely thing to be doing.

File: crypto/ocsp/ocsp_vfy.c (hunk 2 above)
Comment (paulidale) [reply crypto/ocsp/ocsp_vfy.c:210]: Okay, did bother.

File: crypto/evp/p5_crpt2.c (hunk 1 above)
Comment (paulidale) [reply crypto/evp/p5_crpt2.c:24]: Possibly, the work around is fairly trivial: get the digest name.

File: crypto/pem/pvkfmt.c (hunk 4 above)
Comment (paulidale) [reply crypto/pem/pvkfmt.c:798]: It's guarded by a `#ifndef OPENSSL_NO_RC4`.  I'm not seeing a compelling need to introduce an _ex flavour.
Moreover, none of its callers have access to a libctx.

It could be pulled out as a standalone KDF I guess.  I'm not seeing much value to doing so.

File: crypto/srp/srp_vfy.c (hunk 5 above)
Comment (paulidale) [reply crypto/srp/srp_vfy.c:576]: Hmmmm, no comment about the `EVP_MD_fetch(NULL, ..., NULL)` two lines down :laughing:  I can't see anywhere to extract a libctx from, so I'm with @t8m, live and let die.

File: crypto/x509/t_x509.c (hunk 6 above)
Comment (paulidale) [reply crypto/x509/t_x509.c:233]: Added FWIW.

File: providers/implementations/kdfs/scrypt.c (hunk 7 above)
Comment (paulidale) [reply providers/implementations/kdfs/scrypt.c:39]: Prefetching is better here.  I've reverted this big of code and reworked the bits in p5_crpt2.c

**Comment by slontis** [comment]
I am never sure if any of the fetches need the fallback path for engines...

**Comment by paulidale** [comment]
Ones that obviously might, I didn't change.  Those with a libctx and properties are most likely safe.  The ones I'd be concerned about are the ones that subtly might depend -- these we can revert as they are bought up.

**Review by t8m — APPROVED** [review]

**Comment by paulidale** [comment]
Merged to master
